### ISSUE TITLE
Fluentbit is stuck, flb chunks are not flushed and not sending data to ElasticSearch

### ISSUE THREAD

**Issue by joshisumit** [fluent/fluent-bit#5217]
## Bug Report

**Describe the bug**
Fluent is sending data to elasticsearch and it works fine for sometime and then its stuck and not sending data to elastic.
On checking fluentbit pod logs, Briefly I see below 3 issues (Attached full logs below):

```
- [ info] [task] re-schedule retry=0x7f6a1ecc7b68 135 in the next 8 seconds
- [ warn] [engine] failed to flush chunk '1-1648141166.449162875.flb', retry in 8 seconds: task_id=517, input=storage_backlog.1 > output=es.0 (out_id=0)
- [debug] [input:tail:tail.0] 0 new files found on path '/var/log/kube-apiserver-audit.log'  
- (even though the file '/var/log/kube-apiserver-audit.log' exists inside a pod )
```

Referred similar issues, still no luck:
-  As per https://github.com/fluent/fluent-bit/issues/3581#issuecomment-1081107592
I have tried to use 1.9.1-debug version as well, still seeing same results
- https://github.com/fluent/fluent-bit/issues/4336

**To Reproduce**
- start fluentbit as Daemonset on K8S and add node affinity to run only on master node
- Use tail plugin to collect API server Audit Logs (Generated on path `/var/log/kube-apiserver-audit.log` )
- Please refer configs mentioned below in Configuration section

After that Seeing below logs frequently in fluentbit pod:

```
[2022/03/30 10:18:38] [debug] [storage] tail.0:1-1648351188.130831321.flb mapped OK
[2022/03/30 10:18:38] [ info] [input:storage_backlog:storage_backlog.1] queueing tail.0:1-1648351188.130831321.flb
[2022/03/30 10:18:38] [ info] [task] re-schedule retry=0x7f6a1ecc7b68 135 in the next 8 seconds
[2022/03/30 10:18:38] [ info] [task] re-schedule retry=0x7f6a1ecd7478 191 in the next 10 seconds
[2022/03/30 10:18:38] [ info] [task] re-schedule retry=0x7f6a1ecb0b48 104 in the next 10 seconds
[2022/03/30 10:18:38] [ info] [task] re-schedule retry=0x7f6a1ed02588 85 in the next 7 seconds
[2022/03/30 10:18:38] [ info] [task] re-schedule retry=0x7f6a1ed02498 64 in the next 11 seconds
[2022/03/30 10:18:38] [ info] [task] re-schedule retry=0x7f6a1ed022e0 30 in the next 11 seconds
[2022/03/30 10:18:38] [ info] [task] re-schedule retry=0x7f6a1eccae08 147 in the next 9 seconds
[2022/03/30 10:18:38] [ info] [task] re-schedule retry=0x7f6a1ecb6c28 28 in the next 9 seconds

....

[2022/03/30 10:18:38] [debug] [input:tail:tail.0] scanning path /var/log/kube-apiserver-audit.log
[2022/03/30 10:18:38] [debug] [input:tail:tail.0] scan_blog add(): dismissed: /var/log/kube-apiserver-audit.log, inode 1882
[2022/03/30 10:18:38] [debug] [input:tail:tail.0] 0 new files found on path '/var/log/kube-apiserver-audit.log'
[2022/03/30 10:18:38] [debug] [input:tail:tail.0] inode=1882 events: IN_MODIFY 
[2022/03/30 10:18:38] [debug] [storage] tail.0:1-1648635501.574187623.flb mapped OK
[2022/03/30 10:18:38] [debug] [storage] [cio file] alloc_size from 1114112 to 1245184
[2022/03/30 10:18:38] [debug] [storage] [cio file] synced at: tail.0/1-1648635501.574187623.flb
[2022/03/30 10:18:38] [debug] [input chunk] update output instances with new chunk size diff=128565
[2022/03/30 10:18:38] [debug] [retry] new retry created for task_id=517 attempts=1
[2022/03/30 10:18:38] [ warn] [engine] failed to flush chunk '1-1648141166.449162875.flb', retry in 8 seconds: task_id=517, input=storage_backlog.1 > output=es.0 (out_id=0)
[2022/03/30 10:18:38] [debug] [storage] tail.0:1-1648351224.176425521.flb mapped OK
[2022/03/30 10:18:38] [ info] [input:storage_backlog:storage_backlog.1] queueing tail.0:1-1648351224.176425521.flb
[2022/03/30 10:18:38] [debug] [storage] tail.0:1-1648351264.88207341.flb mapped OK
[2022/03/30 10:18:38] [ info] [input:storage_backlog:storage_backlog.1] queueing tail.0:1-1648351264.88207341.flb
[2022/03/30 10:18:38] [debug] [storage] tail.0:1-1648351301.408233326.flb mapped OK
```

On checking further inside fluentbit pod, I see chunks are not cleared and filling up the disk:

```
root@fluent-bit-audit-drbxj:/var/log# du -sh flb-storage/tail.0/
26G     flb-storage/tail.0/ <------

root@fluent-bit-audit-qbkw8:/var/log# ls -la kube-apiserver-audit*
-rw------- 1 root root 104857010 Mar 31 08:19 kube-apiserver-audit-2022-03-31T08-19-39.831.log
-rw------- 1 root root  95179025 Mar 31 08:46 kube-apiserver-audit.log
```

**Expected behavior**
- Keep sending data to ElasticSearch, should not be stuck
- Should be able to flush chunk files(`*.flb` files) regularly (in my case directory is `/var/log/flb-storage/tail.0/`

**Your Environment**

* Version used: `1.8.15` and `1.9.1-debug` (tested on both)
* Configuration (fluentbit configuration):
```
  # Configuration files: server, input, filters and output
  # ======================================================
  fluent-bit.conf: |
    [SERVICE]
        Flush         1
        Log_Level     debug
        Daemon        off
        Parsers_File  parsers.conf
        HTTP_Server   On
        HTTP_Listen   0.0.0.0
        HTTP_Port     2020
        storage.path              /var/log/flb-storage/
        storage.sync              normal
        storage.checksum          off
        storage.backlog.mem_limit 50M
    @INCLUDE input-audit.conf
    @INCLUDE output-elasticsearch.conf

  input-audit.conf: |
    [INPUT]
        Name              tail
        Tag               kube-audit.*
        Path              /var/log/kube-apiserver-audit.log
        Parser            json
        DB                /var/log/flb_kube_audit.db
        storage.type      filesystem
        Mem_Buf_Limit     800MB
        Skip_Long_Lines   On
        Refresh_Interval  10
        Buffer_Chunk_Size 4096k
        Buffer_Max_Size   9000k

  output-elasticsearch.conf: |
    [OUTPUT]
        Name            es
        Match           kube-audit.*
        Host            ${KUBE_ELASTICSEARCH_HOST}
        Port            ${KUBE_ELASTICSEARCH_PORT}
        Index           kube-audit-dev
        Buffer_Size     False
        Generate_ID     on
        AWS_Auth        On
        AWS_Region      us-east-1
        Replace_Dots    On
        Retry_Limit     False
        tls             On
        Trace_Error     On
        Logstash_Format   On
        Logstash_Prefix   kube-audit-dev

  parsers.conf: |
    [PARSER]
        Name   json
        Format json
        Time_Key time
        Time_Format %d/%b/%Y:%H:%M:%S %z

    [PARSER]
        Name        docker
        Format      json
        Time_Key    time
        Time_Format %Y-%m-%dT%H:%M:%S.%L
        Time_Keep   On
```

* Environment name and version (e.g. Kubernetes? What version?): Kubernetes 1.22.4
* Server type and version: Running as docker image on Kubernetes master node (using docker image: `1.9.1-debug`
* Operating System and version: Linux
* Filters and plugins: Plugin: `tail, es`

**Additional context**
- I am running a separate copy of fluentbit as daemonset (which runs only on master node) and collect API server Audit Logs (Generated on path `/var/log/kube-apiserver-audit.log` )
- Initially, its working fine. After some time, chunks are piling up (failed to flush) and I see retry messages in log and fluentbit stops sending data to ElasticSearch.

**Comment by joshisumit** [comment]
@lecaros @edsiper 

Need your inputs on the same. Let me know If I am missing anything/not configured correctly on Fluentbit configuration.

**Comment by dyipon** [comment]
Same issues with fluent-bit:1.9.2 .

**Comment by lecaros** [comment]
Hi @joshisumit,
do you see the input records metric still going up? (i.e. it continues reading events)
With debug log_level enabled, we should see more details on what's happening with your chunks. Can you share the full log or its tail (while it's not sending any more events to es) at least?
Thanks

**Comment by lecaros** [comment]
Thanks @dyipon for confirming.
Can you @joshisumit confirm you still have this issue?

**Comment by fpytloun** [comment]
I am still seeing this issue with forward output and disk buffering with fluentbit 1.9.3.
```
[2022/05/04 19:15:58] [ info] [task] re-schedule retry=0x7fe1729c8220 7 in the next 4 seconds
[2022/05/04 19:16:02] [ info] [task] re-schedule retry=0x7fe1729c8220 7 in the next 6 seconds
[2022/05/04 19:16:08] [ info] [task] re-schedule retry=0x7fe1729c8220 7 in the next 6 seconds
[2022/05/04 19:16:14] [ info] [task] re-schedule retry=0x7fe1729c8220 7 in the next 6 seconds
[2022/05/04 19:16:20] [ info] [task] re-schedule retry=0x7fe1729c8220 7 in the next 7 seconds
[2022/05/04 19:16:27] [ info] [task] re-schedule retry=0x7fe1729c8220 7 in the next 4 seconds
[2022/05/04 19:16:31] [ info] [task] re-schedule retry=0x7fe1729c8220 7 in the next 6 seconds
[2022/05/04 19:16:37] [ info] [task] re-schedule retry=0x7fe1729c8220 7 in the next 4 seconds
[2022/05/04 19:16:41] [ info] [task] re-schedule retry=0x7fe1729c8220 7 in the next 6 seconds
[2022/05/04 19:16:47] [ info] [task] re-schedule retry=0x7fe1729c8220 7 in the next 4 seconds
[2022/05/04 19:16:51] [ info] [task] re-schedule retry=0x7fe1729c8220 7 in the next 7 seconds
```

**Comment by fpytloun** [comment]
Poor man's workaround might be to delete stucked chunks, eg. periodically delete everything not modified in last hour:

```
find /var/log/flb-storage -type f -name "*.flb" -mmin +60 | xargs rm -f || true
```

**Comment by lecaros** [comment]
@fpytloun do you have repro to this issue?
The messages you posted don't necessarily mean you are having an issue.

Thanks

**Comment by fpytloun** [comment]
Unfortunately, I was not able to find out what's causing this. I just know that it's persistent issue, so I assumed it could be eg. some garbage inside buffered messages?

**Comment by lecaros** [comment]
Hi everyone,
in order to address this, we need a reproducible scenario. If any of you have repro steps and can answer other questions posted to troubleshoot this, please provide them and we'll gladly assist.

**Comment by koosw** [comment]
Hi I got same issue, and resolved the problem by deleting *.flb.

To explain my situation in detail, I was using aws-for-fluent-bit version 2.10.0 (fluent-bit 1.6.x) and the crash continued to occur, so I upgraded to 2.25.1 (fluent-bit 1.9.3).
I wanted to use the stable version of aws-for-fluent-bit, but in the stable version, an error such as invalid back log continued to occur, so I upgraded to the 1.9.x version that can skip the error.

And my fluent-bit situation is like the picture below. The pod with the error has unusually high memory(6xx ~ 8xx)
![Screen Shot 2022-06-01 at 1 09 58 AM](https://user-images.githubusercontent.com/3181841/171258989-dc531854-ed26-44aa-9348-b077b40d822f.png)

And the number of flb files is unusually high. (hundred thousands ~ miliions)
![Screen Shot 2022-06-01 at 1 08 40 AM](https://user-images.githubusercontent.com/3181841/171259376-e5f716f2-4e2a-46db-ba81-198755deb3ee.png)

The reason why so many flb files could not be flushed is guessing that the crash occurred in 1.6.x.
Therefore, it seems to be a problem that occurs when there are many flb files.

**Comment by fpytloun** [comment]
Seems that this deadlock might be caused due to non-unicode characters in log (eg. UTF-16/UTF-32). I tried to ensure these characters will not be present and so far it seems that issue went away.

Beside ensuring on log source side, I also tried this Lua filter to replace non-ascii characters:

```
function deep_replace(obj, source, target)
   if obj then
      if type(obj) == "table" then
         local result = {}
         for k,v in pairs(obj) do
            result[k] = deep_replace(v, source, target)
         end
         return result
      elseif type(obj) == "string" then
         return obj:gsub(source, target)
      else
         return obj
      end
   end

   return obj
end


function escape_non_ascii(tag, timestamp, record)
   local new_record = deep_replace(record, "([\128-\255])", "\\\\%1")
   return 1, timestamp, new_record
end
```

```
    [FILTER]
        Name    lua
        Match   *
        Script  /fluent-bit/scripts/filters.lua
        Call    escape_non_ascii
```

**Comment by awangptc** [comment]
> Hi everyone, in order to address this, we need a reproducible scenario. If any of you have repro steps and can answer other questions posted to troubleshoot this, please provide them and we'll gladly assist.

One of the problems I have with this is that it occurs only when elastic has issues.  And we have not been able to figure out how to force the issue.  It could run for days or weeks without seeing the problem, and when elastic has issues that causes rejections this occurs.  It's difficult to run fluent-bit with debug log enabled for the length of time necessary to reproduce this.  It does appear to occur when elastic prevents ingestion for some significant period of time (at least a few hours).  Is there anything that you can tell us about how fluent-bit handles retrying flb chunks?  It's interesting that the workarounds above of deleting older flb files seems to unblock things.  Is there anything from that that makes sense for us to gather more targetted debugging?

**Comment by awangptc** [comment]
I'm having a hard time collecting the logs when this occurs with debug on.  The issue is, when the error condition occurs, our fluent-bit logs are no longer being sent to elatic.  But the sheer volume of debug logs causes our k8s nodes to rotate the logs more quickly than I can pull them.  I'm sure their in flb files but I don't know how to extract them.

The scenario that causes the problem for us is, something cause elastic search to start reject requests - we have seen two conditions that cause this
1) excessive load resulting in elastic sending back 429s to requests
2) elastic ran out of disk space

when these occur, fluent-bit stops sending logs to elastic but continues to read from tail into new flb files.

If we restart fluent-bit it attempts to reprocess the flb files, but if there are multiple nodes, restarting all of the fluent-bit pods will very very quickly overwhelm elastic adn elastic will start to throttle..  429s.  

This is resulting in a high retry/failure rate.  We have our retry limit set to 10.  If a limit is reached what happens to the flb files?  Do they get removed for that transaction?  If not, that might be part of our issue.  I've discovered that if i rotate each fluent-bit pod ONE at a time, elastic does not get overloaded and it is able to work through the backlog flb files.  But that's a time consuming process of course.

**Comment by allenhaozi** [comment]
met same error, but I use loki as backend 
https://github.com/fluent/fluent-bit/issues/6670

**Comment by aslafy-z** [comment]
I have the same issue with fluentd as a backend (using logging-operator)

**Comment by lecaros** [comment]
Hello @awangptc,
> If a limit is reached what happens to the flb files?

flb files are no longer retried and are deleted from the storage.

Are you all using the latest image? We've recently (last month) merged some fixes related to irrecoverable chunks handling.

**Comment by lecaros** [comment]
Please try the latest image. Either 2.0.8 or even nightly builds.

**Comment by rectalogic** [comment]
Having a similar issue with public.ecr.aws/aws-observability/aws-for-fluent-bit:init-2.31.2 sending logs to AWS OpenSearch.

```
[2023/02/28 20:41:40] [ warn] [engine] failed to flush chunk '6-1677616889.860927077.flb', retry in 21 seconds: task_id=2, input=forward.1 > output=opensearch.1 (out_id=1)
[2023/02/28 20:41:41] [ warn] [engine] chunk '6-1677616877.850448169.flb' cannot be retried: task_id=1, input=forward.1 > output=opensearch.1
[2023/02/28 20:41:42] [ warn] [engine] failed to flush chunk '6-1677616901.873936214.flb', retry in 10 seconds: task_id=1, input=forward.1 > output=opensearch.1 (out_id=1)
[2023/02/28 20:41:42] [ warn] [engine] failed to flush chunk '6-1677616901.874024444.flb', retry in 6 seconds: task_id=4, input=forward.1 > output=opensearch.1 (out_id=1)
[2023/02/28 20:41:43] [ warn] [engine] chunk '6-1677616877.850342703.flb' cannot be retried: task_id=0, input=forward.1 > output=opensearch.1
[2023/02/28 20:41:48] [ warn] [engine] failed to flush chunk '6-1677616901.874024444.flb', retry in 7 seconds: task_id=4, input=forward.1 > output=opensearch.1 (out_id=1)
```

**Comment by awangptc** [comment]
We've been running fluent-bit 2.0.8 for about a month now and things seem more stable.  We haven't had a catastropihic outage of elastic that seems to trigger the rproblem before (elastic being down for hours at a time) but we have had elastic being down for 20-30 minutes or so at a time and so far we haven't seen fluent-bit 2.0.8 get stuck.  We're planning on trying to implement a failure situation - shutdown elastic, or block it's port from the client - for a few hours at a time to see if the issue still occurs but it's a little lower on our priority list to try.

**Comment by devanshbaffle** [comment]
@awangptc would you mind sharing your fluent-bit config? I am running into issues with logs not being forwarded. Only 1 pod (base nginx test pod) I can find being forwarded in Kibana. I cannot see my application logs, I am not sure why. I get various different issues no matter what I set. Here is my config.

```
apiVersion: v1
items:
- apiVersion: v1
  data:
    custom_parsers.conf: |
      [PARSER]
          Name docker_no_time
          Format json
          Time_Keep On
          Time_Key time
          Time_Format %Y-%m-%dT%H:%M:%S.%L
    fluent-bit.conf: |
      [SERVICE]
          Daemon Off
          Flush 1
          Log_Level info
          Parsers_File parsers.conf
          Parsers_File custom_parsers.conf
          HTTP_Server On
          HTTP_Listen 0.0.0.0
          HTTP_Port 2020
          Health_Check On

      [INPUT]
          Name tail
          Path /var/log/containers/*.log
          multiline.parser docker, cri
          Tag kube.*
          Mem_Buf_Limit 50MB
          Skip_Long_Lines Off

      [INPUT]
          Name systemd
          Tag host.*
          Systemd_Filter _SYSTEMD_UNIT=kubelet.service
          Read_From_Tail On

      [FILTER]
          Name kubernetes
          Match kube.*
          Merge_Log On
          Keep_Log Off
          K8S-Logging.Parser On
          K8S-Logging.Exclude On

      [OUTPUT]
          Name es
          Match kube.*
          Host            ${FLUENT_ELASTICSEARCH_HOST}
          Port            ${FLUENT_ELASTICSEARCH_PORT}
          HTTP_User       ${FLUENT_ELASTICSEARCH_USER}
          HTTP_Passwd     ${FLUENT_ELASTICSEARCH_PASSWD}
          tls             On
          tls.verify      On
          tls.ca_file     ${FLUENT_ELASTICSEARCH_TLSCA}
          tls.crt_file    ${FLUENT_ELASTICSEARCH_TLSCRT}
          tls.key_file    ${FLUENT_ELASTICSEARCH_TLSKEY}
          Logstash_Format On
          Suppress_Type_Name On
          Retry_Limit     False
          
      [OUTPUT]
          Name es
          Match host.*
          Host            ${FLUENT_ELASTICSEARCH_HOST}
          Port            ${FLUENT_ELASTICSEARCH_PORT}
          HTTP_User       ${FLUENT_ELASTICSEARCH_USER}
          HTTP_Passwd     ${FLUENT_ELASTICSEARCH_PASSWD}
          tls             On
          tls.verify      On
          tls.ca_file     ${FLUENT_ELASTICSEARCH_TLSCA}
          tls.crt_file    ${FLUENT_ELASTICSEARCH_TLSCRT}
          tls.key_file    ${FLUENT_ELASTICSEARCH_TLSKEY}
          Logstash_Format On
          Suppress_Type_Name On
          Logstash_Prefix node
          Retry_Limit False
```

**Comment by rectalogic** [comment]
I am having this issue with the AWS fluentbit fork - I filed a bug with them that has a full reproducible example with all configs here https://github.com/aws/aws-for-fluent-bit/issues/568

I also tested those configs with fluent/fluent-bit:2.0.9 and was able to reproduce it.

In both cases, if I comment out the two "parser" filters in my `filters.conf`, then this issue does not happen (but of course my messages are unparsed)

**Comment by devanshbaffle** [comment]
I was able to fix my issue with Replace_Dots On in my config file. Apparently, it is required since Elasticsearch 2.0.x, but I am running the latest Elasticsearch 8.5.x.

I hope this helps someone.

**Comment by rectalogic** [comment]
I tried with `Replace_Dots On` in the opensearch output but with the same results. 2 of the messages make it to opensearch (like before, but now with dots replaced) and 2 still fail with "failed to flush chunk"

**Comment by leonardo-albertovich** [comment]
@rectalogic could you put together a minimal reproduction case? If you do please ping me and I'll try to fix it.

**Comment by rectalogic** [comment]
@leonardo-albertovich see https://github.com/aws/aws-for-fluent-bit/issues/568 - it has all the config files and the docker command to run to reproduce. You just need to change the `-e OPENSEARCH_FQDN=search.play-test.internal.aws.cureatr.com` on that docker command to point to your OpenSearch instance.

Also if you change the image in that command from `public.ecr.aws/aws-observability/aws-for-fluent-bit:init-2.31.3` to `fluent/fluent-bit:2.0.9` you can still reproduce the issue.

**Comment by leonardo-albertovich** [comment]
Thank you, I'll take a look at it as soon as I can and get back to you.

**Comment by rectalogic** [comment]
Hmm, I noticed if I do *not* pass `-v` (debug) to fluent-bit, then only one log makes it into OpenSearch, the other 3 fail. When I do enable debug, then 2 logs make it to OpenSearch.

e.g. `sudo docker run --rm -e OPENSEARCH_FQDN=search.play-test.internal.aws.cureatr.com -v $PWD/fluent-debug.conf:/fluent-bit/etc/fluent-bit.conf -v $PWD:/fluent-bit-configs/ fluent/fluent-bit:2.0.9 /fluent-bit/bin/fluent-bit -c /fluent-bit/etc/fluent-bit.conf -R /fluent-bit-configs/parsers.conf`

This results in output:
```
Fluent Bit v2.0.9
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2023/03/09 15:06:56] [ info] [fluent bit] version=2.0.9, commit=4c0ca4fc5f, pid=1
[2023/03/09 15:06:56] [ info] [storage] ver=1.4.0, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2023/03/09 15:06:56] [ info] [cmetrics] version=0.5.8
[2023/03/09 15:06:56] [ info] [ctraces ] version=0.2.7
[2023/03/09 15:06:56] [ info] [input:dummy:dummy.0] initializing
[2023/03/09 15:06:56] [ info] [input:dummy:dummy.0] storage_strategy='memory' (memory only)
[2023/03/09 15:06:56] [ info] [input:dummy:dummy.1] initializing
[2023/03/09 15:06:56] [ info] [input:dummy:dummy.1] storage_strategy='memory' (memory only)
[2023/03/09 15:06:56] [ info] [input:dummy:dummy.2] initializing
[2023/03/09 15:06:56] [ info] [input:dummy:dummy.2] storage_strategy='memory' (memory only)
[2023/03/09 15:06:56] [ info] [input:dummy:dummy.3] initializing
[2023/03/09 15:06:56] [ info] [input:dummy:dummy.3] storage_strategy='memory' (memory only)
[2023/03/09 15:06:56] [ info] [output:stdout:stdout.1] worker #0 started
[2023/03/09 15:06:56] [ info] [sp] stream processor started
[0] nginx-firelens: [1677521897.000000000, {"ecs.version"=>"1.8.0", "source.ip"=>"10.1.0.40", "http.request.method"=>"GET", "http.request.referrer"=>"", "http.response.status_code"=>200, "http.response.body.bytes"=>16, "user.id"=>"", "url.original"=>"/", "user_agent.original"=>"ELB-HealthChecker/2.0", "trace.id"=>"FLUENTDEBUG"}]
[0] nginx-firelens: [1678374416.575859503, {"message"=>"FLUENTDEBUG"}]
[0] api-firelens: [1678374416.575881880, {"service.name"=>"uwsgi", "service.type"=>"request", "http.request.method"=>"GET", "url.path"=>"/", "user_agent"=>"ELB-HealthChecker/2.0", "uwsgi.duration.ms"=>3, "http.request.referrer"=>"-", "http.response.status_code"=>200, "url.domain"=>"10.1.1.138:5001", "uwsgi.ctime"=>"Tue Feb 28 17:33:33 2023", "source.ip"=>"10.1.0.188", "trace.id"=>"FLUENTDEBUG", "user.id"=>"-"}]
[0] api-firelens: [1677521897.000000000, {"log.level"=>"info", "message"=>"Processing /", "ecs"=>{"version"=>"1.6.0"}, "event"=>{"action"=>"root", "duration"=>4128932}, "log"=>{"logger"=>"cureatr.url", "origin"=>{"file"=>{"line"=>83, "name"=>"middleware.py"}, "function"=>"tween"}}, "process"=>{"name"=>"MainProcess", "pid"=>15, "thread"=>{"id"=>140134088726336, "name"=>"uWSGIWorker1Core0"}}, "service.name"=>"uwsgi", "source.ip"=>"10.1.0.188", "trace.id"=>"FLUENTDEBUG"}]
[2023/03/09 15:06:57] [ warn] [engine] failed to flush chunk '1-1678374416.575913047.flb', retry in 8 seconds: task_id=3, input=dummy.3 > output=opensearch.0 (out_id=0)
[2023/03/09 15:06:57] [ warn] [engine] failed to flush chunk '1-1678374416.575758816.flb', retry in 6 seconds: task_id=0, input=dummy.0 > output=opensearch.0 (out_id=0)
[2023/03/09 15:06:57] [ warn] [engine] failed to flush chunk '1-1678374416.575883047.flb', retry in 8 seconds: task_id=2, input=dummy.2 > output=opensearch.0 (out_id=0)
[2023/03/09 15:07:03] [ warn] [engine] failed to flush chunk '1-1678374416.575758816.flb', retry in 8 seconds: task_id=0, input=dummy.0 > output=opensearch.0 (out_id=0)
[2023/03/09 15:07:05] [ warn] [engine] failed to flush chunk '1-1678374416.575913047.flb', retry in 17 seconds: task_id=3, input=dummy.3 > output=opensearch.0 (out_id=0)
[2023/03/09 15:07:05] [ warn] [engine] failed to flush chunk '1-1678374416.575883047.flb', retry in 16 seconds: task_id=2, input=dummy.2 > output=opensearch.0 (out_id=0)
[2023/03/09 15:07:11] [error] [engine] chunk '1-1678374416.575758816.flb' cannot be retried: task_id=0, input=dummy.0 > output=opensearch.0
^C[2023/03/09 15:07:16] [engine] caught signal (SIGINT)
[2023/03/09 15:07:16] [ warn] [engine] service will shutdown in max 5 seconds
```

You can see the "failed to flush chunk" messages are for tasks 0, 2 and 3 - task 1 made it. The message that made it was the one that the parser filters did not apply to.

**Comment by leonardo-albertovich** [comment]
That's interesting, I'll try to get to it soon, I'm sorry but I have absolutely no clue of what could be happening in this case so I'll only be able to give you some feedback once I get to reproduce it.

**Comment by leonardo-albertovich** [comment]
That thing about the parser filter sounds promising though

**Comment by rectalogic** [comment]
If I comment out both parser filters, then all 4 messages make it (but of course they are unparsed so not useful)

**Comment by leonardo-albertovich** [comment]
Have you tried changing the output plugin with `stdout` to see what the records look like?

**Comment by rectalogic** [comment]
Yes, the configs I linked to have both `opensearch` and `stdout` outputs. The records look correct - you can see them in the logs I posted above.

**Comment by leonardo-albertovich** [comment]
Sorry for the obvious question, as I said I haven't been able to dedicate the time and attention required to this and was just glancing and asking questions to see if there was anything I could do without switching context.

**Comment by rectalogic** [comment]
I enabled `trace_output On` and `trace_error On` in my `opensearch` output and can now see the issue. My opensearch instance must already have mappings for some of these and the fluent-bit parsers are generating documents that don't match those mappings.

```
[2023/03/20 19:50:00] [debug] [output:opensearch:opensearch.0] HTTP Status=200 URI=/_bulk
[2023/03/20 19:50:00] [debug] [output:opensearch:opensearch.0] error caused by: Input
{"create":{"_index":"filebeat-ecs-2023.03.20","_type":"_doc"}}
{"@timestamp":"2023-03-20T19:49:43.575Z","service.name":"uwsgi","service.type":"request","http.request.method":"GET","url.path":"/","user_agent":"ELB-HealthChecker/2.0","uwsgi.duration.ms":3,"http.request.referrer":"-","http.response.status_code":200,"url.domain":"10.1.1.138:5001","uwsgi.ctime":"Tue Feb 28 17:33:33 2023","source.ip":"10.1.0.188","trace.id":"FLUENTDEBUG","user.id":"-"}


[2023/03/20 19:50:00] [error] [output:opensearch:opensearch.0] error: Output
{"took":7,"errors":true,"items":[{"create":{"_index":"filebeat-ecs-2023.03.20","_type":"_doc","_id":"xGqSAIcBswdmZijlTAkK","status":400,"error":{"type":"mapper_parsing_exception","reason":"Could not dynamically add mapping for field [source.ip]. Existing mapping for [source] must be of type object but found [text]."}}}]}
```

**Comment by leonardo-albertovich** [comment]
I googled the error and saw some references but if I'm honest I don't know anything about it. Have you tried modifying your payload so instead of defining "source.ip" as a string you define it as a dict this way? `"source": {"ip": "value"}`

Excuse me if that's a stupid question but I'm interested in how this turns out.

**Comment by leonardo-albertovich** [comment]
Btw, do you think you could put together a simple repro case for that?

**Comment by rectalogic** [comment]
The issue is conflicting mappings, during earlier experiments I think text got uploaded as `source` and now it's trying to treat `source` as an object due to `source.ip`.

You should be able to repro using the configs I provided and change the dummy inputs so that they generate conflicting dynamic mappings, e.g.
```
[INPUT]
  Name dummy
  Tag api-firelens
  Samples 1
  Dummy {"log": "{\"foobar\": 123}"}

[INPUT]
  Name dummy
  Tag api-firelens
  Samples 1
  Dummy {"log": "{\"foobar\": \"text\"}"}
```

So this will first map "foobar" as integer, then send text which should fail.

**Comment by leonardo-albertovich** [comment]
Ok, then this is a case where logging should be improved to be a bit more meaningful rather than a bug in fluent-bit right?

I wonder what would be a good middle ground between dumping the response and parsing / printing the message on error each time (Some users might not like it if it changes the default behavior and if it doesn't then there's not much difference with having to toggle the option  to dump the response body on error...).

Any ideas?

**Comment by leonardo-albertovich** [comment]
I encourage you to submit an improvement proposal with the expected behavior from the users point of view, I think it would be very good. I'm not familiar with that plugin or system so I can't say much but I'm all in for improving the user experience and would really appreciate the help.

**Comment by awangptc** [comment]
the issue for us has not been resolved with 2.0.8.  So what I haven't figured out is how to reliably reproduce this.
What happens is
1) something causes elastic to stop accepting ingestion
2) fluent bit buffers to storage unable to send to elastic
3) over some period of time ... this is what i haven't been able to determine
4) we fix the elastic problem
5) fluent-bit is stuck in "retry" 

last night elastic was stuck in read only for about 5 hours due to filesystem full (non critical system so we don't have any paging weekend/overnight expectations)

After fixing teh elastic problem, existing fluent-bit pods were all stuck on retry with no output records being processed.
After restarting fluent-bit it will work through but one of the issues is, our fluent-bit throughput is considerably faster than elastic can process (6000+records per second) which results in elastic throttling (http 429 resonse codes).  This can also eventually get fluent-bit stuck as it's unable to send requests ( see step 3 above - i have not been able to identify a trigger, is it time based?  some other criteria?) and the only fix i've found is to restart the fluent-bit pods

The symptoms for fluent bit are stuck with logging like this
```
[2023/03/26 17:36:59] [ info] [task] re-schedule retry=0x7f41f0bce430 1204 in the next 1 seconds
[2023/03/26 17:36:59] [ info] [task] re-schedule retry=0x7f41f0bcf4a0 1205 in the next 1 seconds
```

**Comment by awangptc** [comment]
> The issue is conflicting mappings, during earlier experiments I think text got uploaded as `source` and now it's trying to treat `source` as an object due to `source.ip`.
> 
> You should be able to repro using the configs I provided and change the dummy inputs so that they generate conflicting dynamic mappings, e.g.
> 
> ```
> [INPUT]
>   Name dummy
>   Tag api-firelens
>   Samples 1
>   Dummy {"log": "{\"foobar\": 123}"}
> 
> [INPUT]
>   Name dummy
>   Tag api-firelens
>   Samples 1
>   Dummy {"log": "{\"foobar\": \"text\"}"}
> ```
> 
> So this will first map "foobar" as integer, then send text which should fail.

Mappings conflict is one potential cause of this issue, but it is not the only one.  If you set your fluent-bit to a limited number of retries (we set it to 10) the mapping conflicts eventually get dropped and they are not going to perpetually retry.  In that scenario, the case I described where elastic stops accepting ingestion, still occurs.

**Comment by fpytloun** [comment]
> This issue is stale because it has been open 90 days with no activity. Remove stale label or comment or this will be closed in 5 days. Maintainers can add the `exempt-stale` label.

not stalled

**Comment by aslafy-z** [comment]
Still a thing with fluentbit 2.1.10

**Comment by lecaros** [comment]
@aslafy-z do you have repro steps? That would be really helpful.

**Comment by carlosrmendes** [comment]
facing the same with v2.2.1, any feedback @awangptc ?

**Comment by rafael-gumiero** [comment]
I also have the same issue in the version 2.2.2, any feedback @awangptc ?

**Comment by saurabhgcect** [comment]
@awangptc : Any solution for above,currently using Fluentbit 2.2.2

**Comment by smartbenson1** [comment]
same issue here....

**Comment by Bishal-Bhandari01** [comment]
same issue here on version 2.2.2 as shown in the @rectalogic image while adding output as an syslog.

`
[OUTPUT]
        Name syslog
        Match *
        Host <host ip-address>
        Port 514
        Mode udp
        # syslog_maxsize       2048
        # syslog_format        rfc5424
        # syslog_severity_key  severity
        # syslog_facility_key  facility
        # syslog_hostname_key  hostname
        # syslog_appname_key   appname
        # syslog_procid_key    procid
        # syslog_msgid_key     msgid
        # syslog_sd_key        sd
        # syslog_message_key   message
`

Steps to reproduce:
- add the above code in the values.yaml and comment down the other outputs.
- replace this code:
`
hostNetwork: false
dnsPolicy: ClusterFirst
`
to
`
hostNetwork: true
dnsPolicy: ClusterFirstWithHostNet
`
- rename the file as fluent-bit-values.yaml
- create a namespace called logging in k3s
- install fluent-bit using helm
`
helm upgrade --install logging fluent/fluent-bit --values ./fluent-bit-values.yaml
`

@lecaros can you help me with problem

**Comment by itsobgraph** [comment]
Also seeing this. Using aws-for-fluent-bit version 2.31.12.20231011

**Comment by Venture200** [comment]
I can see that there are chunks from 2, 3 even 4 days old. Wondering why fluent-bit doesnt delete them automatically instead keep them forever in one specific node ?

**Comment by daihengbin724** [comment]
> Hello, I am having similar issue, in one of my kubernetes nodes, fluent-bit persistent chunk size is increasing significantly around 15GB while other nodes dont have such issue. I am using 3.2.1

same issue

**Comment by Venture200** [comment]
> > Hello, I am having similar issue, in one of my kubernetes nodes, fluent-bit persistent chunk size is increasing significantly around 15GB while other nodes dont have such issue. I am using 3.2.1
> 
> same issue

I have created separate ticket, please mention there as well [9669](https://github.com/fluent/fluent-bit/issues/9669#issuecomment-2507712879)
